bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,347 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    hash,
+    local::{salt::Salt, Local},
+    peer::{peer_id::PeerId, Peer},
+};
+
+use prost::bytes::{Buf, Bytes};
+
+use std::{
+    cmp, fmt,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+    vec,
+};
+
+/// The distance between the local entity and a neighbor.
+pub type Distance = u32;
+
+pub(crate) const SIZE_INBOUND: usize = 4;
+pub(crate) const SIZE_OUTBOUND: usize = 4;
+
+/// Decides whether a peer is a suitable neighbor, or not.
+pub trait NeighborValidator
+where
+    Self: Send + Sync + Clone,
+{
+    /// Returns `true` if the given [`Peer`](crate::peer::Peer) is a valid neighbor.
+    fn is_valid(&self, peer: &Peer) -> bool;
+}
+
+// A neighbor is a peer with a distance) metric.
+#[derive(Clone, Debug)]
+pub(crate) struct Neighbor {
+    peer: Peer,
+    distance: Distance,
+}
+
+impl Neighbor {
+    pub(crate) fn new(peer: Peer, distance: Distance) -> Self {
+        Self { peer, distance }
+    }
+
+    pub(crate) fn peer(&self) -> &Peer {
+        &self.peer
+    }
+
+    pub(crate) fn distance(&self) -> Distance {
+        self.distance
+    }
+}
+
+impl fmt::Display for Neighbor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.peer().peer_id(), self.distance())
+    }
+}
+
+impl Eq for Neighbor {}
+impl PartialEq for Neighbor {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
```

Comment:
What does that mean exactly ? Two different peers at equal distance are equal neighbors ?

Reply:
yes....for the peering algorithm the concrete peer behind a neighbor is irrelevant. All it cares about is to find the N closest peers. Let's say there's just one open slot, but there are two candidates with exactly the same distance (but further away than all existing neighbors), then it can choose 1 of them randomly. While `Peer`s are compared by their `PeerId`, neighbors are only compared by their `Distance`.